zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/message/MessageCorrelationProcessor.java

```diff
@@ -0,0 +1,132 @@
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.engine.processing.message;
+
+import io.camunda.zeebe.engine.processing.bpmn.behavior.BpmnBehaviors;
+import io.camunda.zeebe.engine.processing.common.EventHandle;
+import io.camunda.zeebe.engine.processing.message.command.SubscriptionCommandSender;
+import io.camunda.zeebe.engine.processing.streamprocessor.TypedRecordProcessor;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.StateWriter;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.TypedResponseWriter;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.Writers;
+import io.camunda.zeebe.engine.state.immutable.EventScopeInstanceState;
+import io.camunda.zeebe.engine.state.immutable.MessageStartEventSubscriptionState;
+import io.camunda.zeebe.engine.state.immutable.MessageState;
+import io.camunda.zeebe.engine.state.immutable.MessageSubscriptionState;
+import io.camunda.zeebe.engine.state.immutable.ProcessState;
+import io.camunda.zeebe.protocol.impl.record.value.message.MessageCorrelationRecord;
+import io.camunda.zeebe.protocol.impl.record.value.message.MessageRecord;
+import io.camunda.zeebe.protocol.record.intent.MessageCorrelationIntent;
+import io.camunda.zeebe.protocol.record.intent.MessageIntent;
+import io.camunda.zeebe.stream.api.records.TypedRecord;
+import io.camunda.zeebe.stream.api.state.KeyGenerator;
+
+public final class MessageCorrelationProcessor
+    implements TypedRecordProcessor<MessageCorrelationRecord> {
+
+  private final MessageCorrelateBehavior correlateBehavior;
+  private final KeyGenerator keyGenerator;
+  private final StateWriter stateWriter;
+  private final TypedResponseWriter responseWriter;
+
+  public MessageCorrelationProcessor(
+      final Writers writers,
+      final KeyGenerator keyGenerator,
+      final EventScopeInstanceState eventScopeInstanceState,
+      final ProcessState processState,
+      final BpmnBehaviors bpmnBehaviors,
+      final MessageStartEventSubscriptionState startEventSubscriptionState,
+      final MessageState messageState,
+      final MessageSubscriptionState messageSubscriptionState,
+      final SubscriptionCommandSender commandSender) {
+    stateWriter = writers.state();
+    responseWriter = writers.response();
+    this.keyGenerator = keyGenerator;
+    final var eventHandle =
+        new EventHandle(
+            keyGenerator,
+            eventScopeInstanceState,
+            writers,
+            processState,
+            bpmnBehaviors.eventTriggerBehavior(),
+            bpmnBehaviors.stateBehavior());
+    correlateBehavior =
+        new MessageCorrelateBehavior(
+            startEventSubscriptionState,
+            messageState,
+            eventHandle,
+            stateWriter,
+            messageSubscriptionState,
+            commandSender);
+  }
+
+  @Override
+  public void processRecord(final TypedRecord<MessageCorrelationRecord> command) {
```

Comment:
💭 I think the processor should be as simple as possible. All the logic should be moved into the Behaviour class. In this way the processor is in charge to only call the right methods, and all the logic will be in the same place. Wdyt ?

Reply:
Sorry maybe my explanation is not good enough. What I want to say is that we can move all the logic in the `processRecord()` method into the behaviour class, but NOT in the already existing methods.

At the end the processor should be like:

```
@Override
public void processRecord(final TypedRecord<MessageCorrelationRecord> command) {
    correlateBehaviour.doSomething(.....)
}
```